diff.go

```diff
@@ -0,0 +1,146 @@
+// be send to Ralph. Other fields provide convenience shourtcuts for some
+// functions/methods (e.g. Component field holds a reference to the original
+// object, which frees us from unmarshaling contents of Data field).
+type DiffComponent struct {
+	ID        int       // ID of the object held in Component field
+	Name      string    // name of the component (e.g. Ethernet, Memory)
+	Data      []byte    // JSON-ed Component field
+	Component Component // reference to the original object
+}
+
+// NewDiffComponent creates a DiffComponent based on a given component. Since
+// component is an interface type, it can hold both object or pointer, but
+// NewDiffComponent handles both of these cases.
+func NewDiffComponent(component Component) (*DiffComponent, error) {
+	var id int
+	var name string
+	var data []byte
+	var err error
+	switch v := component.(type) {
+	case *Ethernet:
+		id = v.ID
+		name = "Ethernet"
+		data, err = json.Marshal(v)
+		if err != nil {
+			return nil, err
+		}
+	case Ethernet:
+		return NewDiffComponent(&v)
+	case *Memory:
+		id = v.ID
+		name = "Memory"
+		data, err = json.Marshal(v)
+		if err != nil {
+			return nil, err
+		}
+	case Memory:
+		return NewDiffComponent(&v)
+	default:
+		return nil, fmt.Errorf("unknown component: %+v", v)
+	}
+	return &DiffComponent{
+		ID:        id,
+		Name:      name,
+		Data:      data,
+		Component: component,
+	}, nil
+}
+
+func (d DiffComponent) String() string {
+	return fmt.Sprintf("DiffComponent{ID: %d, Name: %s, Data: %s, Component: %s}",
+		d.ID, d.Name, string(d.Data), d.Component)
+}
+
+// SendDiffToRalph sends a given Diff to Ralph. If dryRun is set to true, then
+// no changes will be sent to Ralph. If noOutput is set to true, then all the
+// output from this function will be silenced (this is mostly for tests).
+// Returned statusCodes slice is meant only to facilitate tests, so don't be
+// surprised if you see it ignored somewhere in the source code.
+func SendDiffToRalph(client *Client, diff *Diff, dryRun bool, noOutput bool) (statusCodes []int, err error) {
+
+	var send = func(d *DiffComponent, method, endpoint, msg string) (int, error) {
+		var code int
+		var data []byte
+		switch {
+		case method == "DELETE":
+			data = nil
+		default:
+			data = d.Data
+		}
+		if !dryRun {
+			code, err = client.SendToRalph(method, endpoint, data)
+		}
+		if err != nil {
+			return code, err
+		}
+		if !noOutput {
+			fmt.Printf("%s %s successfully.\n", d.Component, msg) // TODO(xor-xor): Use logger instead.
+		}
+		return code, nil
+	}
+
+	for _, d := range diff.Create {
+		endpoint := APIEndpoints[d.Name]
+		code, err := send(d, "POST", endpoint, "created")
+		if err != nil {
+			return statusCodes, err
+		}
+		if code != 0 {
+			statusCodes = append(statusCodes, code)
+		}
+	}
+	for _, d := range diff.Update {
+		endpoint := fmt.Sprintf("%s/%d", APIEndpoints[d.Name], d.ID)
+		code, err := send(d, "PUT", endpoint, "updated")
```

Comment:
As GUIDO says - practicality beats purity. And the author of article is going to the very sad, old, SOAP place with the proposing of strict: 

<img width="710" alt="zrzut ekranu 2016-07-07 o 15 45 02" src="https://cloud.githubusercontent.com/assets/552398/16655279/e13bf238-4459-11e6-8ff4-8bdd2c0d870f.png">

instead of expected REST-ful

<img width="496" alt="zrzut ekranu 2016-07-07 o 15 44 55" src="https://cloud.githubusercontent.com/assets/552398/16655282/e60dfa54-4459-11e6-862d-0780c9763526.png">